ecosystem/indexer-grpc/indexer-grpc-file-store-backfiller/src/processor.rs

```diff
@@ -0,0 +1,387 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    process::exit,
+    sync::Arc,
+    time::Duration,
+};
+use tokio::sync::Mutex;
+
+/// Processor tails the data in cache and stores the data in file store.
+pub struct Processor {
+    file_store_operator: Box<dyn FileStoreOperator>,
+    chain_id: u64,
+    grpc_stream: Option<tonic::Streaming<TransactionsFromNodeResponse>>,
+    starting_version: u64,
+    progress_file_path: String,
+    ending_version: Option<u64>,
+    validation_mode: bool,
+    backfill_processing_task_count: usize,
+    validating_task_count: usize,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ProgressFile {
+    version: u64,
+}
+
+impl Processor {
+    pub async fn new(
+        fullnode_grpc_address: url::Url,
+        file_store_config: IndexerGrpcFileStoreConfig,
+        chain_id: u64,
+        enable_cache_compression: bool,
+        progress_file_path: String,
+        starting_version: Option<u64>,
+        transactions_count: Option<u64>,
+        validation_mode: bool,
+        backfill_processing_task_count: usize,
+        validating_task_count: usize,
+    ) -> Result<Self> {
+        let _cache_storage_format = if enable_cache_compression {
+            StorageFormat::Lz4CompressedProto
+        } else {
+            StorageFormat::Base64UncompressedProto
+        };
+        if validation_mode {
+            return Ok(Self {
+                file_store_operator: file_store_config.create(),
+                chain_id,
+                grpc_stream: None,
+                starting_version: starting_version.unwrap_or(0),
+                progress_file_path,
+                ending_version: transactions_count.map(|c| starting_version.unwrap_or(0) + c),
+                validation_mode,
+                backfill_processing_task_count,
+                validating_task_count,
+            });
+        }
+
+        let starting_version = starting_version.unwrap_or(0);
+        let expected_end_version = transactions_count.map(|c| starting_version + c);
+        // Resume from the progress file if the file exists. If not, create an empty one.
+        let progress_file: ProgressFile = match std::fs::read(&progress_file_path) {
+            Ok(bytes) => serde_json::from_slice(&bytes).context("Failed to parse progress file")?,
+            Err(_) => {
+                let progress_file = ProgressFile {
+                    version: starting_version,
+                };
+                let bytes = serde_json::to_vec(&progress_file)
+                    .context("Failed to serialize progress file")?;
+                std::fs::write(&progress_file_path, bytes)
+                    .context("Failed to write progress file")?;
+                progress_file
+            },
+        };
+        let expected_starting_version = std::cmp::max(starting_version, progress_file.version);
+        tracing::info!(
+            "Starting backfill from version {}",
+            expected_starting_version
+        );
```

Comment:
nit: use the starting_version = expected_starting_version for easier finding on humio?

Reply:
done.